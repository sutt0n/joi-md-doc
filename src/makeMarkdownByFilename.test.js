const joi = require('@hapi/joi');
const makeMarkdownByFilename = require('./makeMarkdownByFilename');
const R = require('ramda');

describe(`makeMarkdownByFilename`, () => {
  describe(`unit`, () => {
    it(`should support primitive type`, () => {
      const joiSchema = {
        describe: () => ({
          metas: [{ name: 'Test', filename: 'test' }],
          keys: {
            foo: { type: 'number' },
            bar: { type: 'string' },
            baz: { type: 'boolean' },
            baz: { type: 'date' },
          },
        }),
      };

      expect(makeMarkdownByFilename(joiSchema)).toMatchSnapshot();
    });

    it(`should support nested schemas/objects`, () => {
      const joiSchema = {
        describe: () => ({
          metas: [{ name: 'Parent', filename: 'parent' }],
          keys: {
            foo: {
              type: 'object',
              metas: [{ name: 'Child', filename: 'child' }],
              keys: {
                foo: { type: 'number' },
              },
            },
          },
        }),
      };

      expect(makeMarkdownByFilename(joiSchema)).toMatchSnapshot();
    });

    it(`should support nested schemas/objects with no name meta tag`, () => {
      const joiSchema = {
        describe: () => ({
          metas: [{ filename: 'parent' }],
          keys: {
            foo: {
              type: 'object',
              metas: [{ filename: 'child' }],
              keys: {
                foo: { type: 'number' },
              },
            },
          },
        }),
      };

      expect(makeMarkdownByFilename(joiSchema)).toMatchSnapshot();
    });

    it(`should support list of single primitive type`, () => {
      const joiSchema = {
        describe: () => ({
          metas: [{ name: 'Test', filename: 'test' }],
          keys: {
            stringList: {
              type: 'array',
              items: [{ type: 'string' }],
            },
          },
        }),
      };

      expect(makeMarkdownByFilename(joiSchema)).toMatchSnapshot();
    });

    it(`should support list of various primitive types`, () => {
      const joiSchema = {
        describe: () => ({
          metas: [{ name: 'Test', filename: 'test' }],
          keys: {
            stringList: {
              type: 'array',
              items: [
                { type: 'string' },
                { type: 'number' },
                { type: 'boolean' },
                { type: 'date' },
              ],
            },
          },
        }),
      };

      expect(makeMarkdownByFilename(joiSchema)).toMatchSnapshot();
    });

    it(`should support nested list of single schema/object type`, () => {
      const joiSchema = {
        describe: () => ({
          metas: [{ name: 'Test', filename: 'test' }],
          keys: {
            stringList: {
              type: 'array',
              items: [
                {
                  type: 'object',
                  metas: [{ name: 'List Item', filename: 'listItem' }],
                  keys: { listItemKey: { type: 'string' } },
                },
              ],
            },
          },
        }),
      };

      expect(makeMarkdownByFilename(joiSchema)).toMatchSnapshot();
    });

    it(`should mark required fields`, () => {
      const joiSchema = {
        describe: () => ({
          metas: [{ name: 'Test', filename: 'test' }],
          keys: {
            foo: {
              type: 'number',
              flags: {
                presence: 'required',
              },
            },
          },
        }),
      };

      expect(makeMarkdownByFilename(joiSchema)).toMatchSnapshot();
    });

    it(`should support descriptions`, () => {
      const joiSchema = {
        describe: () => ({
          metas: [{ name: 'Test', filename: 'test' }],
          keys: {
            foo: {
              type: 'number',
              flags: {
                description: 'This is a description.',
              },
            },
          },
        }),
      };

      expect(makeMarkdownByFilename(joiSchema)).toMatchSnapshot();
    });

    it(`should support default meta tags`, () => {
      const joiSchema = {
        describe: () => ({
          metas: [{ name: 'Test', filename: 'test' }],
          keys: {
            foo: {
              type: 'number',
              metas: [{ default: 1 }],
            },
          },
        }),
      };

      expect(makeMarkdownByFilename(joiSchema)).toMatchSnapshot();
    });

    it(`should support whitelisting valid values`, () => {
      const joiSchema = {
        describe: () => ({
          metas: [{ name: 'Test', filename: 'test' }],
          keys: {
            foo: {
              type: 'string',
              allow: ['valid1', 'valid2'],
            },
          },
        }),
      };

      expect(makeMarkdownByFilename(joiSchema)).toMatchSnapshot();
    });

    it(`should not show whitelist when list length >= maxValueListLength`, () => {
      const joiSchema = {
        describe: () => ({
          metas: [{ name: 'Test', filename: 'test' }],
          keys: {
            foo: {
              type: 'string',
              allow: R.range(0, 100),
            },
          },
        }),
      };

      expect(makeMarkdownByFilename(joiSchema)).toMatchSnapshot();
    });

    it(`should support blacklisting invalid values`, () => {
      const joiSchema = {
        describe: () => ({
          metas: [{ name: 'Test', filename: 'test' }],
          keys: {
            foo: {
              type: 'string',
              invalid: ['invalid1', 'invalid2'],
            },
          },
        }),
      };

      expect(makeMarkdownByFilename(joiSchema)).toMatchSnapshot();
    });

    it(`should not show blacklist when list length >= maxValueListLength`, () => {
      const joiSchema = {
        describe: () => ({
          metas: [{ name: 'Test', filename: 'test' }],
          keys: {
            foo: {
              type: 'string',
              invalid: R.range(0, 100),
            },
          },
        }),
      };

      expect(makeMarkdownByFilename(joiSchema)).toMatchSnapshot();
    });

    it(`should support min and max`, () => {
      const joiSchema = {
        describe: () => ({
          metas: [{ name: 'Test', filename: 'test' }],
          keys: {
            foo: {
              type: 'number',
              rules: [
                {
                  name: 'min',
                  args: {
                    limit: 0,
                  },
                },
                {
                  name: 'max',
                  args: {
                    limit: 5,
                  },
                },
              ],
            },
          },
        }),
      };

      expect(makeMarkdownByFilename(joiSchema)).toMatchSnapshot();
    });

    it(`should support length`, () => {
      const joiSchema = {
        describe: () => ({
          metas: [{ name: 'Test', filename: 'test' }],
          keys: {
            foo: {
              type: 'array',
              rules: [
                {
                  name: 'length',
                  args: {
                    limit: 1,
                  },
                },
              ],
            },
          },
        }),
      };

      expect(makeMarkdownByFilename(joiSchema)).toMatchSnapshot();
    });

    it(`should use "then" branch when "alternatives" type`, () => {
      const joiSchema = {
        describe: () => ({
          metas: [{ name: 'Test', filename: 'test' }],
          keys: {
            foo: {
              type: 'alternatives',
              matches: [
                {
                  then: {
                    type: 'string',
                  },
                },
              ],
            },
          },
        }),
      };

      expect(makeMarkdownByFilename(joiSchema)).toMatchSnapshot();
    });

    it(`should support "any" types`, () => {
      const joiSchema = {
        describe: () => ({
          metas: [{ name: 'Test', filename: 'test' }],
          keys: {
            foo: {
              type: 'any',
            },
          },
        }),
      };

      expect(makeMarkdownByFilename(joiSchema)).toMatchSnapshot();
    });

    it(`negative: should throw when no filename meta tag`, () => {
      const joiSchema = {
        describe: () => ({
          metas: [{ name: 'Test' }],
        }),
      };

      expect(() => makeMarkdownByFilename(joiSchema)).toThrowError(
        'Required filename meta tag not provided.',
      );
    });

    it(`should default name to filename`, () => {
      const joiSchema = {
        describe: () => ({
          metas: [{ filename: 'test' }],
          keys: {
            foo: { type: 'number' },
          },
        }),
      };

      expect(makeMarkdownByFilename(joiSchema)).toMatchSnapshot();
    });

    it(`should exclude key when isDocumented meta tag is false`, () => {
      const joiSchema = {
        describe: () => ({
          metas: [{ name: 'Test', filename: 'test' }],
          keys: {
            included: { type: 'string' },
            excluded: { type: 'string', metas: [{ isDocumented: false }] },
          },
        }),
      };

      expect(makeMarkdownByFilename(joiSchema)).toMatchSnapshot();
    });
  });

  describe(`integration`, () => {
    it(`should support appended schemas`, () => {
      const baseSchema = joi.object().keys({
        base: joi.boolean(),
      });

      const extendedSchema = baseSchema
        .append({
          extended: joi.boolean(),
        })
        .meta({
          name: 'Extends from Base',
          filename: 'extendedFromBase',
        });

      expect(makeMarkdownByFilename(extendedSchema)).toMatchSnapshot();
    });
  });
});
