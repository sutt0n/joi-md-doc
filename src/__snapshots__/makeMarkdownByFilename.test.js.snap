// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeMarkdownByFilename integration should support appended schemas 1`] = `
Object {
  "extendedFromBase": "# Extends from Base

## base

> \`boolean\` | optional

## extended

> \`boolean\` | optional
",
}
`;

exports[`makeMarkdownByFilename unit should default name to filename 1`] = `
Object {
  "test": "# test

## foo

> \`number\` | optional
",
}
`;

exports[`makeMarkdownByFilename unit should exclude key when isDocumented meta tag is false 1`] = `
Object {
  "test": "# Test

## included

> \`string\` | optional
",
}
`;

exports[`makeMarkdownByFilename unit should mark required fields 1`] = `
Object {
  "test": "# Test

## foo

> \`number\` | required
",
}
`;

exports[`makeMarkdownByFilename unit should not show blacklist when list length >= maxValueListLength 1`] = `
Object {
  "test": "# Test

## foo

> \`string\` | optional
",
}
`;

exports[`makeMarkdownByFilename unit should not show whitelist when list length >= maxValueListLength 1`] = `
Object {
  "test": "# Test

## foo

> \`string\` | optional
",
}
`;

exports[`makeMarkdownByFilename unit should support "any" types 1`] = `
Object {
  "test": "# Test

## foo

> \`*\` | optional
",
}
`;

exports[`makeMarkdownByFilename unit should support blacklisting invalid values 1`] = `
Object {
  "test": "# Test

## foo

> \`string\` | optional


blacklist: \`invalid1\`, \`invalid2\`
",
}
`;

exports[`makeMarkdownByFilename unit should support default meta tags 1`] = `
Object {
  "test": "# Test

## foo

> \`number\` | optional | defaults to \`1\`
",
}
`;

exports[`makeMarkdownByFilename unit should support descriptions 1`] = `
Object {
  "test": "# Test

## foo

> \`number\` | optional


This is a description.
",
}
`;

exports[`makeMarkdownByFilename unit should support length 1`] = `
Object {
  "test": "# Test

## foo

> \`array\` | optional | length: \`1\`
",
}
`;

exports[`makeMarkdownByFilename unit should support list of single primitive type 1`] = `
Object {
  "test": "# Test

## stringList

> \`array\`: \`string\` | optional
",
}
`;

exports[`makeMarkdownByFilename unit should support list of various primitive types 1`] = `
Object {
  "test": "# Test

## stringList

> \`array\`: \`string\`, \`number\`, \`boolean\`, \`date\` | optional
",
}
`;

exports[`makeMarkdownByFilename unit should support min and max 1`] = `
Object {
  "test": "# Test

## foo

> \`number\` | optional | min: \`0\` | max: \`5\`
",
}
`;

exports[`makeMarkdownByFilename unit should support nested list of single schema/object type 1`] = `
Object {
  "listItem": "[Test](./test.md)
# List Item

## listItemKey

> \`string\` | optional
",
  "test": "# Test

## stringList

> \`array\`: [List Item](./listItem.md) | optional
",
}
`;

exports[`makeMarkdownByFilename unit should support nested schemas/objects 1`] = `
Object {
  "child": "[Parent](./parent.md)
# Child

## foo

> \`number\` | optional
",
  "parent": "# Parent

## [foo](./child.md)

> \`object\` | optional
",
}
`;

exports[`makeMarkdownByFilename unit should support nested schemas/objects with no name meta tag 1`] = `
Object {
  "child": "[parent](./parent.md)
# child

## foo

> \`number\` | optional
",
  "parent": "# parent

## [foo](./child.md)

> \`object\` | optional
",
}
`;

exports[`makeMarkdownByFilename unit should support primitive type 1`] = `
Object {
  "test": "# Test

## foo

> \`number\` | optional

## bar

> \`string\` | optional

## baz

> \`boolean\` | optional

## faz

> \`date\` | optional
",
}
`;

exports[`makeMarkdownByFilename unit should support whitelisting valid values 1`] = `
Object {
  "test": "# Test

## foo

> \`string\` | optional


whitelist: \`valid1\`, \`valid2\`
",
}
`;

exports[`makeMarkdownByFilename unit should use "then" branch when "alternatives" type 1`] = `
Object {
  "test": "# Test

## foo

> \`string\` | optional
",
}
`;
